libsolidity/analysis/ControlFlowGraph.h

```diff
@@ -0,0 +1,140 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/ast/ASTVisitor.h>
+#include <libsolidity/interface/ErrorReporter.h>
+
+#include <map>
+#include <memory>
+#include <stack>
+#include <vector>
+
+namespace dev
+{
+namespace solidity
+{
+
+/// Basic Control Flow Block.
+/// Basic block of control flow. Consists of a set of AST nodes
+/// for which control flow is always linear.
+struct ControlFlowBlock
+{
+	/// All variable declarations executed in this control flow block.
+	std::vector<VariableDeclaration const*> variableDeclarations;
+	/// All expressions executed in this control flow block (this includes all subexpressions!).
+	std::vector<Expression const*> expressions;
+	/// All inline assembly statements executed in this control flow block.
+	std::vector<InlineAssembly const*> inlineAssemblyStatements;
+	/// If control flow returns in this node, the return statement is stored in returnStatement,
+	/// otherwise returnStatement is nullptr.
+	Return const* returnStatement = nullptr;
+};
+
+/// Node of the Control Flow Graph.
+/// The control flow is a directed graph connecting control flow blocks.
+/// An arc between two nodes indicates that the control flow can possibly
+/// move from its start node to its end node during execution.
+struct CFGNode
+{
+	/// Entry nodes. All CFG nodes from which control flow may move into this node.
+	std::vector<CFGNode*> entries;
+	/// Exit nodes. All CFG nodes to which control flow may continue after this node.
+	std::vector<CFGNode*> exits;
+
+	/// Control flow in the node.
+	ControlFlowBlock block;
+};
+
+/// Describes the control flow of a function.
+struct FunctionFlow
+{
+	FunctionFlow(CFGNode* _entry, CFGNode* _exit, CFGNode* _exception):
+		entry(_entry), exit(_exit), exception(_exception) {}
+	/// Entry node. Control flow of the function starts here.
+	/// This node does not have any entries.
+	CFGNode* entry;
```

Comment:
This might be confusing when compared with CFGNode. Here, it designates the first node of the function, while in CFGNode it designates the predecessors.

Reply:
In both cases it's the node from which control flow enters - so I would say it's exactly the same as in CFGNode and hence it makes sense to use the same name.